keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry/README.md

```diff
@@ -0,0 +1,593 @@
+registry for faster development.
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- Users may see the ConfigMap and draw mistaken conclusions about how to
+  interact with it. They think deleting the ConfigMap would delete the local
+  registry (which it does not).
+
+- This KEP does not specify any automatic mechanism for keeping the
+  LocalRegistryStatus up-to-date with how the cluster is configured.
+  For example, the user might delete the registry. The cluster might
+  not have a way of knowing that the registry has died.
+  
+- This KEP creates the new API version `local-registry.x-k8s.io/v1alpha1`,
+  This may open a can of worms. On the one hand, cluster implementations want
+  confidence that they're implementing a shared standard. On the other hand,
+  a new API version implies standards of governance, ownership, and API machinery
+  that might not make sense yet.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+None
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Tools that configure a local registry should apply the following ConfigMap to
+the cluster:
+
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: local-image-loaders
+  namespace: kube-public
+data:
+  registryStatus.v1: |
+    host: "localhost:5000"
+    hostFromCluster: "registry:5000"
+    help: "https://kind.sigs.k8s.io/docs/user/local-registry/"
+```
+
+Documentation that educates people on how to set up a registry manually should
+include instructions on how this is applied.
+
+The name of the config map must be `local-image-loaders`.
+
+The namespace must be `kube-public`.
+
+The `registryStatus.v1` field of the ConfigMap data must contain YAML. 
+The following code is the reference implementation that readers may use to parse the YAML:
+
+```go
+
+// A local registry allows clients to load images into the local cluster by
+// pushing to this registry.
+type LocalRegistryStatusV1 struct {
+	// Host documents the host (hostname and port) of the registry, as seen from
+	// outside the cluster.
+	//
+	// This is the registry host that tools should push images to.
+	Host string `yaml:"host,omitempty"`
```

Comment:
a slightly delayed question (my fault), but how do we deal with the case where the cluster admin does not want the world to see the registry or at least what is our recommendation?

do the users only set `HostFromCluster` in such a case?

Reply:
I think the cluster admin would delete all the contents of localRegistryStatus.v1

I added some comments to hostFromCluster to make it clear that it should be ignored if host is not set

Reply:
> I added some comments to hostFromCluster to make it clear that it should be ignored if host is not set

a cluster admin might want the registry host to be visible locally by setting "hostFromCluster" (assuming it would contain a local address), but they might not want to set "host" as this would be visible to the world.

isn't it a valid use case to only set "hostFromCluster" but leave "host" empty?

Reply:
ah, I think there might be a misunderstanding of what this is for.

`hostFromCluster` isn't for in-cluster tools that want to discover the registry

it's for out-of-cluster tools that need to send things to in-cluster tools.

it's pretty subtle and not super ergonomic, though easier for cluster maintainers to implement

Reply:
I think the confusion here speaks to a legitimate ambiguity in the API - the distinction between  network addresses from the point of view of the nodes & pods, and network addresses from the point of view of the container runtime (i.e., which has some additional configuration and mirroring). There was a good discussion about this in the kind slack channel today. 

We think it might make sense to break this up into two fields, but i want to think a bit about how to name these fields.

Reply:
> I think the confusion here speaks to a legitimate ambiguity in the API 

yes, i think i agree.

also...naming is hard, we usually bikeshed on written proposals in Google docs and Zoom meetings before sending a KEP PR, so i blame myself for not recommending some of that here too.